Apollo federation docs (#42)
mirumee/ariadne-website: docs/apollo-federation.md

---
id: apollo-federation
title: Apollo Federation
---

> [Apollo Federation](https://www.apollographql.com/docs/apollo-server/federation/introduction/) is an architecture that composes multiple GraphQL services into a single data graph. Ariadne supports building federated schemas that rely on [additional types and directives](https://www.apollographql.com/docs/apollo-server/federation/federation-spec/) to instrument the behaviour of the underlying graph and convey the relationships between schema types.

Reviewer: The wording here feels really technical on folk without super-good english - this may hurt search engine and seo.

How about:

```suggestion
[Apollo Federation](https://www.apollographql.com/docs/apollo-server/federation/introduction/) is an approach for composing multiple GraphQL services into a single data graph, queryable from single GraphQL server.

Ariadne supports building federated schemas through use of special [types and directives introduced by federation specification](https://www.apollographql.com/docs/apollo-server/federation/federation-spec/) that instruct Ariadne how it's GraphQL schema types and fields combine with types and fields defined in other GraphQL schemas forming the Federation.
```

Reviewer: Also we could add this below, for more SEO/docs search juice:

> **Note:** You may have heard about "schema stitching" when searching for GraphQL solutions for merging multiple GraphQL servers into a single one. This approach was originally proposed back in 2018 and implemented by the Apollo Server, but it had considerable downsides which resulted in it being scrapped and federations being proposed as new, better solution.

> [Apollo Federation](https://www.apollographql.com/docs/apollo-server/federation/introduction/) is an architecture of composing multiple GraphQL services into a single data graph. Ariadne supports building the federated schemas that rely on [additional types and directives](https://www.apollographql.com/docs/apollo-server/federation/federation-spec/) to instrument the behavior of the underlying graph and convey the relationships between schema types.

Reviewer 2: ```suggestion
[Apollo Federation](https://www.apollographql.com/docs/apollo-server/federation/introduction/) is an architecture that composes multiple GraphQL services into a single data graph. Ariadne supports building the federated schemas that rely on [additional types and directives](https://www.apollographql.com/docs/apollo-server/federation/federation-spec/) to instrument the behavior of the underlying graph and convey the relationships between schema types.
```

## Federated architecture example

Let's say we have three separate GraphQL services that we want to convert into a federated service:

> Users

Reviewer: ```suggestion
Users:
```

```graphql
type Query {
  me: User
}

type User {
  id: ID!
  name: String
  email: String!
}
```

> Reviews

Reviewer: ```suggestion
Reviews:
```

```graphql
type Query {
  reviews(first: Int = 5): [Review]
}

type Review {
  id: ID!
  body: String
}
```

> Products

Reviewer: ```suggestion
Products service:
```

In our case we should add it to the `User`, `Review` and `Product` types:

```graphql
type User @key(fields: "email") {
  id: ID!
  name: String
  email: String!
}
```

```graphql
type Review @key(fields: "id") {
  id: ID!
  body: String
}
```

```graphql
type Product @key(fields: "upc") {
  upc: String!
  name: String
  price: Int
  weight: Int
}
```

The next step is to connect together our types in the distributed architecture.

Let's extend types that are defined by another implementing service to illustrate the power of the federation. In our case the reviews service extends the `User` and `Product` types by adding a reviews field to them:

```graphql
type Review @key(fields: "id") {
  id: ID!
  body: String
  author: User @provides(fields: "email")
  product: Product @provides(fields: "upc")
}

type User @key(fields: "email") @extends {
  email: String! @external
  reviews: [Review]
}

type Product @key(fields: "upc") @extends {
  upc: String! @external
  reviews: [Review]
}
```

> Now, our federation schemas are ready. It's time for revolvers.

Reviewer: ```suggestion
Now, our federated schemas are ready. It's time for revolvers.
```

type_defs = """
  type Query {
    reviews(first: Int = 5): [Review]
  }

  type Review @key(fields: "id") {
    id: ID!
    body: String
    author: User @provides(fields: "email")
    product: Product @provides(fields: "upc")
  }

  type User @key(fields: "email") @extends {
    email: String! @external
    reviews: [Review]
  }

  type Product @key(fields: "upc") @extends {
    upc: String! @external
    reviews: [Review]
  }
"""

review = FederatedObjectType("Review")
user = FederatedObjectType("User")
product = FederatedObjectType("Product")

@review.reference_resolver
def resolve_reviews_reference(_, _info, representation):
    return get_review_by_id(representation["id"])

@review.field("author")
def resolve_review_author(review, *_):
    return {"email": review["user"]["email"]}

@review.field("product")
def resolve_review_product(review, *_):
    return {"upc": review["product"]["upc"]}

@user.field("reviews")
def resolve_user_reviews(representation, *_):
    return get_user_reviews(representation["email"])

@product.field("reviews")
def resolve_product_reviews(representation, *_):
    return get_product_reviews(representation["upc"])

```

> Finally, we need to use the `make_federated_schema` function in every our service to augment our schema definition with federation support:

Reviewer 2: ```suggestion
Finally, we need to use the `make_federated_schema` function in each of our services to augment the schema definition with federation support:
```
